### ISSUE TITLE
access list (whitelist of allowed client)

### ISSUE THREAD

**Issue by ramarro123** [0xERR0R/blocky#1408]
Hello,

it's possible (will be possible) to add a feature to filter the inbound client to any of the port that blocky expose? including but not limited to prometheus exporter, api and so on?

in my idea something like
allowedClients:
-127.0.0.1
-192.0.0.1/24

not sure if the request is clear enough sorry my english is not the best :)

the scenario i am facing is that i want to use blocky on my box, expose it to public (dns over tls or over http, still to decide)
and use it in my mobile/other devices.

i know that i still have to open to a huge subnet (mobile phone(s) of my provider) but is still an added security.

Another use case is to restrict access to prometheus exporter, to avoid sending sensitive data to everyone know the path

**Comment by kwitsch** [comment]
It's currently not possible through blocky itself. 
Usually best practice would be to use a reverse proxy like traefik to restrict outbound access by setting up restriction rules per path/port.

Only restricting the IPs could be implemented fairly easy but would get tricky and insufficient if blocky gets accessed through a container network.

**Comment by ThinkChaos** [comment]
I agree with @kwitsch best practice would be a reverse proxy if you're using DoH/DoT.  
And for filtering IPs, a firewall is more appropriate IMO.

I think the one thing blocky could do here to make this kind of setup easier to secure is allow setting up DoH separately from other HTTP endpoints. That way metrics, API, etc. can listen on localhost/LAN only and DoH on all interfaces.  
A reverse proxy will already let you achieve this kind of granularity, but it's a lot more complex.

**Comment by gspannu** [comment]
**I think Blocky may/ should still implement a client access/ block list for DoT/DoH queries.**

DoT and DoH users may want to identify the connecting clients connecting to Blocky for DNS resolution, for managing statistics, etc... 

For example, e.g. `https://myblockyaddress.com/dns-query/clientxyz` or `tls://clientxyz.myblockyaddress.com`
These clients are then identified in the query log (by the field ClientName)

A section for client access would be useful (something on the lines of below). 

This may work as either Allowed or Blocked.

Everything **Blocked** except for these IP addresses or Client Names
```
DNSClients:
  Allowed:
    - 100.100.99.99
    - 100.200.100.200
    - 192.168.0.0/16
    - clientxyz  # https Client 1
    - clientabc # tls Client 2
    
```

Everything **Allowed** except for these IP addresses or Client Names
```
DNSClients:
  Blocked:
    - 123.123.123.123
    - 10.9.8.0/24
    - clientpqr # blocked client 1
    
```
 
 
or it could do something like this for Everything **Blocked** except for these IP addresses or Client Names
```
DNSClients:
  Allowed:
    - 192.168.0.0/16
    - 100.200.100.200
    - clientxyz  # https Client 1
    - clientabc # tls Client 2
   Blocked:
    - *
```


AGH support a similar feature where it can control which clients are allowed/blocked from accessing DoH/DoT services.

**Comment by gspannu** [comment]
> I agree with @kwitsch best practice would be a **reverse proxy if you're using DoH/DoT**..
> 

---

Could you please advise as to **how to setup Reverse Proxy for limiting DoH/DoT access to certain clients only**?

My blocky instance is accessible to the outside world (it is hosted on a Virtual Machine, behind a router) as `https://myblockyaddress.com/dns-query/` and also as `tls://myblockyaddress.com`

I would like to restrict access to this Blocky by all clients except for a specific client **client0001**
meaning I only want Blocky to respond to requests sent as DoH/DoT as  
`https://myblockyaddress.com/dns-query/client0001` or `tls://client0001.myblockyaddress.com`. 

All other queries should be blocked. e.g. `https://myblockyaddress.com/dns-query/` or `tls://myblockyaddress.com` or `https://myblockyaddress.com/dns-query/client1` should be blocked.

_I am running Caddy as the Reverse Proxy on my router and I have a wildcard certificate for *.myblockyaddress.com_

- My router (192.168.1.1) is running Caddy
- Blocky is hosted internally on another machine at 192.168.1.254
- Blocky using standard ports for DoT/DoH (853, 443)

What config do I need to put in Caddy to manage this situation of **allowing only a specific client/ clients** access to Blocky DNS resolution.

**Comment by ThinkChaos** [comment]
> I think Blocky may/ should still implement a client access/ block list for DoT/DoH queries.

I understand that as a user it would be nice to setup from blocky config, but it's not worth implementing an inferior solution here compared to what a firewall/reverse proxy can already do.  
One example of an issue is it's much more costly for the requests to arrive to all the way to blocky, and blocky do the required processing to decide the request is not allowed, than for the firewall to drop the packets immediately based on its rules. That lowers the cost of a denial of service from a blocked client.

> I would like to restrict access to this Blocky by all clients except for a specific client **client0001** meaning I only want Blocky to respond to requests sent as DoH/DoT as `https://myblockyaddress.com/dns-query/client0001` or `tls://client0001.myblockyaddress.com`.

This is what reverse proxy are all about: they know about the HTTP path for DoH, and TLS server name for DoT and can decide what to do with each request.  
For Caddy the way to only expose a service at a path/servername seems to be done with [matchers](https://caddyserver.com/docs/caddyfile/matchers).  **This is insecure if not paired with further defense** because anyone can guess/find out about `client0001`.

However you go about it, exposing blocky (or any DNS server) on the internet is **not recommended**, unless you know what you're doing and willing to deal with the fallout.  
**The best way to get access to your home DNS from anywhere is using a VPN**. Generally, the VPN should be the only thing in your home listening on the internet directly as it is designed for that and gives you access to everything else (some other services fit that category, but by default everything should be accessed through your VPN).

TL;DR:
- set up a VPN in your home network and expose it to the internet
- connect your clients to the VPN
- query blocky through the VPN

If you really want IP/hostname based access, use a firewall or reverse proxy.

I made a spinoff issue for making DoH listen on a different port from other HTTP(S) endpoints.

**Comment by gspannu** [comment]
> I made a spinoff issue for making DoH listen on a different port from other HTTP(S) endpoints.

Thank you for this...
